File: iOSObjectiveCFoundation/Classes/Frontend/Component/Vfx/Bouncy.m
// ======================================================================
// Project Name    : ios_foundation
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
#import "Bouncy.h"
static const float kGravity = 0.98f;
static const float kE = 0.5f;
@implementation Bouncy
@synthesize gravity;
@synthesize velocity;
@synthesize originVelocity;
- (id)init:(float)velocity rate:(float)rate {
    self = [super init];
    self.velocity = velocity;
    self.originVelocity = velocity;
    self.gravity = kGravity * rate;
    return self;
}
- (float)update {
    self.velocity -= self.gravity;
    return self.velocity;
}
- (void)restore {
    self.originVelocity *= kE;
    self.velocity = self.originVelocity;
    return;
}
@end

File: iOSSwiftFoundation/Classes/Frontend/Component/Vfx/Bouncy.swift
// ======================================================================
// Project Name    : ios_foundation
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
import Foundation
public class Bouncy {
    fileprivate static let GRAVITY: Float = 0.98
    fileprivate static let E: Float = 0.5
    public var gravity: Float!
    public var velocity: Float!
    public var originVelocity: Float!
    public init(velocity: Float, rate: Float) {
        self.velocity = velocity
        self.originVelocity = velocity
        self.gravity = Bouncy.GRAVITY * rate
    }
    open func update() -> Float {
        self.velocity -= self.gravity
        return self.velocity
    }
    open func restore() -> Void {
        self.originVelocity *= Bouncy.E
        self.velocity = self.originVelocity
        return
    }
}
